Implemented success response with an empty array

JacekAndrzejewski opened the pull request:
1. Pulled out default success response data to a field
2. Added fluent method for setting default success response data
3. Added tests for usage of new method
4. Added missing tests for null values in respondWithSuccess, respondCreated, respondNoContent and fixed problems
    - null values were allowed both in types and phpdocs, yet exceptions were thrown when passed, this is fixed and null values are returned as empty arrays
5. Added missing tests for Arrayable, JsonSerializable and Collection and their empty variants
6. Strengthened tests to use strict assertions

JacekAndrzejewski committed:
Fixed passing null to respondWithSuccess
Added missing test cases

Signed-off-by: JacekAndrzejewski <jacekandrzejewskiit@gmail.com>

JacekAndrzejewski committed:
Stregthened test assertions to use strict comparisons

Signed-off-by: JacekAndrzejewski <jacekandrzejewskiit@gmail.com>

JacekAndrzejewski committed:
Added internal phpdoc tags to methods that shouldn't be called/overwritten outside of trait

Signed-off-by: JacekAndrzejewski <jacekandrzejewskiit@gmail.com>

JacekAndrzejewski committed:
Implemented functionality to set the default success response

Signed-off-by: JacekAndrzejewski <jacekandrzejewskiit@gmail.com>

JacekAndrzejewski committed:
Fixed passing null to respondNoContent and respondCreated, unify default argument values

Signed-off-by: JacekAndrzejewski <jacekandrzejewskiit@gmail.com>

JacekAndrzejewski committed:
Revert "Added internal phpdoc tags to methods that shouldn't be called/overwritten outside of trait"

This reverts commit 0be9e6fc77f1a943ef0366fb736607539b292c29.

JacekAndrzejewski commented on src/ApiResponseHelpers.php:
```diff
@@ -14,6 +14,9 @@
 trait ApiResponseHelpers
 {
+
+    private ?array $_api_helpers_defaultSuccessData = ['success' => true];
```
That name probably won't be used by any project anywhere, but I'm glad to hear suggestions

JacekAndrzejewski commented on src/ApiResponseHelpers.php:
```diff
@@ -33,17 +36,23 @@ public function respondNotFound(
      */
-    public function respondWithSuccess($contents = []): JsonResponse
+    public function respondWithSuccess($contents = null): JsonResponse
     {
-        $contents = $this->morphToArray($contents);
+        $contents = $this->morphToArray($contents) ?? [];
```
Change of default value won't impact calls without arguments, but will allow to cleanly change passed nulls to empty arrays when someone does pass null. Empty arrays will stay the same as before.
This change is in respondNoContent and respondCreated too.

ultrono replied:
Yep kudos there, this was exactly my plan as to avoid a full breaking change 👍

JacekAndrzejewski commented on tests/ResponseTest.php:
```diff
@@ -35,8 +36,21 @@ public function testResponses(string $method, array $args, int $code, array $dat
         self::assertInstanceOf(JsonResponse::class, $response);
-        self::assertEquals($code, $response->getStatusCode());
-        self::assertEquals($data, $response->getData(true));
+        self::assertSame($code, $response->getStatusCode());
```
Ummm yeah, that's not what happened sir Git

JacekAndrzejewski commented on tests/ResponseTest.php:
```diff
@@ -74,6 +88,13 @@ public function basicResponsesDataProvider(): array
             ['success' => true],
           ],
 
+          'respondWithSuccess(), null response data' => [
+            'respondWithSuccess',
+            [null],
+            Response::HTTP_OK,
+            ['success' => true],
+          ],
```
This case was missing and failed before my changes. Null is allowed based on phpdocs and type hints. There are few similar cases down below.

JacekAndrzejewski commented on tests/ResponseTest.php:
```diff
@@ -95,6 +116,69 @@ public function basicResponsesDataProvider(): array
+            ['success' => true],
+          ],
+
+          'respondWithSuccess(), Arrayable' => [
+            'respondWithSuccess',
+            [
+              new class implements Arrayable {
+                public function toArray()
+                {
+                  return ['id' => 1, 'name' => 'John'];
+                }
+              },
+            ],
+            Response::HTTP_OK,
+            ['id' => 1, 'name' => 'John']
+          ],
+
+          'respondWithSuccess(), empty Arrayable' => [
+            'respondWithSuccess',
+            [
+              new class implements Arrayable {
+                public function toArray()
+                {
+                  return [];
+                }
+              },
+            ],
+            Response::HTTP_OK,
+            ['success' => true]
+          ],
```
Those didn't fail before my changes, however if you would break morphToArray on purpose for Arrayable only then that would not be caught by tests. That's why I added those.

JacekAndrzejewski commented on tests/ResponseTest.php:
```diff
@@ -35,8 +36,21 @@ public function testResponses(string $method, array $args, int $code, array $dat
+     * @dataProvider successDefaultsDataProvider
+     * @throws JsonException
+     */
+    public function testSuccessResponseDefaults(?array $default, $args, int $code, array $data): void
+    {
+        $response = $this->service->setDefaultSuccessResponse($default)->respondWithSuccess($args);
+        self::assertInstanceOf(JsonResponse::class, $response);
+        self::assertSame($code, $response->getStatusCode());
+        self::assertSame($data, $response->getData(true));
```
We could technically merge this and other test method, but it would increase complexity too much IMO. It's already complex enough.

JacekAndrzejewski commented on src/ApiResponseHelpers.php:
```diff
@@ -33,17 +36,23 @@ public function respondNotFound(
         return $this->apiResponse($data);
     }
 
+    public function setDefaultSuccessResponse(?array $content = null)
```
We could add a `self` or `static` phpdoc, however from my tests this doesn't help PHPStorm realize that this is fluent method returning class using this trait. I were unable to make it understand that.
But it realizes it returns object with trait methods so it's good enough. If anyone has idea how to make this better then give this a shot.

ultrono replied:
I'll take a look at that, usually `self` has always been enough. However, as this is via a trait, as you've discovered, that may change.

ultrono replied:
PHPStorm was nagging about no return type and suggested `self`. I've added that for good measure.

JacekAndrzejewski commented:
This should solve everything discussed in #13.
It should also be compatible with #15, with possibly minor changes.
It should be backwards compatible, one place where it's not is that passing null no longer throws an exception. If someone called methods with null and used try-catch and ignored exceptions it WOULD be a BC. However I think we can allow that case, because if someone really does that with methods returning responses then they have worse problems with their code.

I tried explaining anything, if something is hard to understand or needs fixing please ask.
@ultrono Please close conversations where you have no questions and run workflows if all is ok.

ultrono commented:
Great PR and explanation @JacekAndrzejewski, much appreciated.  Have finally got my schedule under control and will look at getting this merged asap.

ultrono commented:
After merging this in (everything looks great from an initial look) I think a small update the readme would be beneficial too.

JacekAndrzejewski committed:
Updated README.md

Signed-off-by: JacekAndrzejewski <jacekandrzejewskiit@gmail.com>

ultrono committed:
Optimise imports of main test class

ultrono committed:
Add `self` return type for `setDefaultSuccessResponse()`, update read me wording.

ultrono commented:
@JacekAndrzejewski Thanks for this, great PR 👍

JacekAndrzejewski commented:
@ultrono By the way, could you label this with `hacktoberfest-accepted`? I want to plant a tree :D

ultrono commented:
> 
> 
> @ultrono By the way, could you label this with `hacktoberfest-accepted`? I want to plant a tree :D

No probs, all added.
